### PR TITLE
CASSANDRA-18451: CEP-15: (C*) Improve the chaos generation for Burn Tests: slow/flakey connections and dropped messages

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: accord-core/src/main/java/accord/impl/InMemoryCommandStore.java [hunk 1]
```diff
@@ -398,6 +398,13 @@ protected <T> void executeInContext(InMemoryCommandStore commandStore, PreLoadCo
         }
     }
 
+    @Override
+    public String toString() {
```
Comment (dcapwell) [inline accord-core/src/main/java/accord/impl/InMemoryCommandStore.java]: I use the debugger a lot to make sure threading is correct, so this is for that use case

File: accord-core/src/main/java/accord/messages/Accept.java [hunk 2]
```diff
@@ -77,7 +77,7 @@ private Accept(TxnId txnId, PartialRoute<?> scope, long waitForEpoch, long minEp
     }
 
     @Override
-    public synchronized AcceptReply apply(SafeCommandStore safeStore)
+    public AcceptReply apply(SafeCommandStore safeStore)
```
Comment (dcapwell) [inline accord-core/src/main/java/accord/messages/Accept.java:79]: I couldn't find any shared state, so removed to allow more concurrency; this is unrelated to this patch

File: accord-core/src/main/java/accord/messages/ReadData.java [hunk 3]
```diff
@@ -216,7 +216,7 @@ private void maybeRead(SafeCommandStore safeStore, SafeCommand safeCommand)
                 // nothing to see here
                 break;
             case RETURNED:
-                throw new IllegalStateException("ReadOk was sent, yet ack called again");
+                throw new IllegalStateException("ReadOk was sent, yet ack called again for txn " + txnId);
```
Comment (dcapwell) [inline accord-core/src/main/java/accord/messages/ReadData.java]: was hitting this case, so added to improve debugging

File: accord-core/src/main/java/accord/messages/ReadData.java [hunk 4]
```diff
@@ -267,9 +267,9 @@ private void ack()
             switch (state)
             {
                 case RETURNED:
-                    throw new IllegalStateException("ReadOk was sent, yet ack called again");
+                    throw new IllegalStateException("ReadOk was sent, yet ack called again for txn " + txnId);
                 case OBSOLETE:
-                    logger.debug("After the read completed for txn {}, the result was marked obsolete", txnId);
+                    logger.trace("After the read completed for txn {}, the result was marked obsolete", txnId);
```
Comment (dcapwell) [inline accord-core/src/main/java/accord/messages/ReadData.java]: added this in the last patch, lowering to trace to not spam burn test

File: accord-core/src/test/java/accord/burn/BurnTest.java [hunk 5]
```diff
@@ -217,28 +226,38 @@ static void burn(RandomSource random, TopologyFactory topologyFactory, List<Id>
...
+
+            maybeTriggerNextRequest.run();
 
             try
             {
                 int start = starts[(int)packet.replyId];
                 int end = clock.incrementAndGet();
 
-                logger.debug("{} at [{}, {}]", reply, start, end);
+                logger.debug("{} at [{}, {} (out of {})]", reply, start, end, operations * 2);
```
Comment (dcapwell) [inline accord-core/src/test/java/accord/burn/BurnTest.java]: I find this helpful to track progress

File: accord-core/src/test/java/accord/impl/basic/Cluster.java [hunk 6]
```diff
@@ -132,18 +133,6 @@ private void processNext(Object next)
             Packet deliver = (Packet) next;
             Node on = lookup.apply(deliver.dst);
 
-            // TODO (required, testing): random drop chance independent of partition; also port flaky connections etc. from simulator
-            // Drop the message if it goes across the partition
-            boolean drop = ((Packet) next).src.id >= 0 &&
```
Comment (dcapwell) [inline accord-core/src/test/java/accord/impl/basic/Cluster.java:158]: moved to `accord.impl.basic.NodeSink#maybeEnqueue`

File: accord-core/src/test/java/accord/impl/list/ListRequest.java [hunk 7]
```diff
@@ -108,27 +110,39 @@ public void accept(Result success, Throwable fail)
...
+                switch (s)
+                {
+                    case Invalidated:
+                        node.reply(client, replyContext, ListResult.invalidated(client, ((Packet)replyContext).requestId, txnId));
+                        break;
+                    case Lost:
+                        node.reply(client, replyContext, ListResult.lost(client, ((Packet)replyContext).requestId, txnId));
+                        break;
+                    case Neither:
+                        node.reply(client, replyContext, ListResult.timeout(client, ((Packet)replyContext).requestId, txnId));
```
Comment (dcapwell) [inline accord-core/src/test/java/accord/impl/list/ListRequest.java:142]: might have missed something, but when I hit this in burn the comment `// currently caught elsewhere in response tracking, but might help to throw an exception here` didn't seem to reflect what was going on.

What I saw was that the txn was `Applied`, but the client request `Timeout`, so when we tried to figure out the state it is `Neither` which would not cause us to bump the `clock`, which lead to burn failing due to clock mismatch

File: accord-core/src/test/java/accord/impl/basic/SimulatedDelayedExecutorService.java [hunk 8]
```diff
@@ -18,48 +18,23 @@
...
-
-    private static int microToNanos(int value)
-    {
-        return Math.toIntExact(TimeUnit.MICROSECONDS.toNanos(value));
     }
 
     @Override
     public void execute(Task<?> task)
     {
-        pending.add(task, jitterInNano.getLong(random), TimeUnit.NANOSECONDS);
```
Comment (dcapwell) [inline accord-core/src/test/java/accord/impl/basic/SimulatedDelayedExecutorService.java:62]: jitter was moved to `PendingQueue`

File: accord-core/src/test/java/accord/burn/BurnTest.java [hunk 9]
```diff
@@ -228,35 +234,53 @@ static void burn(RandomSource random, TopologyFactory topologyFactory, List<Id>
...
-            if (replies[(int)packet.replyId] != null)
-                return;
-
+        Runnable maybeTriggerNextRequest = () -> {
             if (requestIndex.get() < requests.length)
             {
                 int i = requestIndex.getAndIncrement();
                 starts[i] = clock.incrementAndGet();
-                queue.add(requests[i]);
+                // TODO (review): why did i add this again?
```
Comment (dcapwell) [inline accord-core/src/test/java/accord/burn/BurnTest.java:242]: Not looked at this since April so I don't remember why I added this... there was a reason, but don't remember...

**Comment by belliottsmith** [comment]
I believe this has been merged already.
